src/Components/Components/src/NavigationManager.cs

```diff
@@ -0,0 +1,220 @@
+        /// Allows derived classes to lazyly self-initialize. Implementations that support lazy-initialization should override
+        /// this method and call <see cref="Initialize(string, string)" />.
+        /// </summary>
+        protected virtual void EnsureInitialized()
+        {
+        }
+
+        /// <summary>
+        /// Converts a relative URI into an absolute one (by resolving it
+        /// relative to the current absolute URI).
+        /// </summary>
+        /// <param name="relativeUri">The relative URI.</param>
+        /// <returns>The absolute URI.</returns>
+        public Uri ToAbsoluteUri(string relativeUri)
+        {
+            AssertInitialized();
+            return new Uri(_baseUri, relativeUri);
+        }
+
+        /// <summary>
+        /// Given a base URI (e.g., one previously returned by <see cref="BaseUri"/>),
+        /// converts an absolute URI into one relative to the base URI prefix.
+        /// </summary>
+        /// <param name="uri">An absolute URI that is within the space of the base URI.</param>
+        /// <returns>A relative URI path.</returns>
+        public string ToBaseRelativePath(string uri)
+        {
+            if (uri.StartsWith(_baseUriString, StringComparison.Ordinal))
+            {
+                // The absolute URI must be of the form "{baseUri}something" (where
+                // baseUri ends with a slash), and from that we return "something"
+                return uri.Substring(_baseUriString.Length);
+            }
+
+            var hashIndex = uri.IndexOf('#');
+            var uriWithoutHash = hashIndex < 0 ? uri : uri.Substring(0, hashIndex);
+            if ($"{uriWithoutHash}/".Equals(_baseUriString, StringComparison.Ordinal))
+            {
+                // Special case: for the base URI "/something/", if you're at
+                // "/something" then treat it as if you were at "/something/" (i.e.,
+                // with the trailing slash). It's a bit ambiguous because we don't know
+                // whether the server would return the same page whether or not the
+                // slash is present, but ASP.NET Core at least does by default when
+                // using PathBase.
+                return uri.Substring(_baseUriString.Length - 1);
+            }
+
+            var message = $"The URI '{uri}' is not contained by the base URI '{_baseUriString}'.";
+            throw new ArgumentException(message);
+        }
+
+        internal static string NormalizeBaseUri(string baseUri)
+        {
+            var lastSlashIndex = baseUri.LastIndexOf('/');
+            if (lastSlashIndex >= 0)
+            {
+                baseUri = baseUri.Substring(0, lastSlashIndex + 1);
+            }
+
+            return baseUri;
+        }
+
+        /// <summary>
+        /// Triggers the <see cref="LocationChanged"/> event with the current URI value.
+        /// </summary>
+        protected void NotifyLocationChanged(bool isInterceptedLink)
```

Comment:
I've been looking around and I haven't seen any place where we use isInterceptedLink (other than to pass it as an argument to LocationChangedEventArgs).

I've also looked for when is `LocationEventChangedEventArgs.IsNavigationIntercepted` is used and is is only used within the router. Seems the reason it is used there is to display a NotFound component in case there is no matching route.

Could we slightly change the router implementation to do the following?
Instead of 
```csharp
if (!isNavigationIntercepted && NotFoundContent != null)
{
    Log.DisplayingNotFoundContent(_logger, locationPath, _baseUri);


    // We did not find a Component that matches the route.
    // Only show the NotFoundContent if the application developer programatically got us here i.e we did not
    // intercept the navigation. In all other cases, force a browser navigation since this could be non-Blazor content.
    _renderHandle.Render(NotFoundContent);
}
else
{
    Log.NavigatingToExternalUri(_logger, _locationAbsolute, locationPath, _baseUri);
    UriHelper.NavigateTo(_locationAbsolute, forceLoad: true);
}
```
* For the first render (Called on SetParameters on the router), always display NotFound component if you can't find a matching component.
* Have the client handle all the routes within the base path instead of selectively using IsNavigationIntercepted. 
  * The client side code already provides an escape hatch to do this by adding `target=_self` to links.
  * The user can specify forceReload: true when navigating programatically to force the page to reload unconditionally.

To put a concrete example, imagine you have a blazor server-side app with Identity in it for auth. The changes you would make are as follows:
* You would add target="_self" in the links to LoginPartial (or equivalent blazor component).
* If you were to have some sort of redirect auth handler, you would write NavigationManager.RedirectTo("/Identity/Account/Login", forceReload: true);

Given these simplifications, I don't think we need the isInterceptedLink concept in the API. We don't need to make the server smart if we make the client explicit, and it also makes it easier to reason about. The common case/pattern implemented by most UI frameworks is that your SPA takes over the url space once it has been loaded on the client.

Reply:
> Have the client handle all the routes within the base path instead of selectively using IsNavigationIntercepted.
> The client side code already provides an escape hatch to do this by adding target=_self to links.
> The user can specify forceReload: true when navigating programatically to force the page to reload unconditionally.

That's how we used to do it, and people didn't like it. People were reporting bugs that they couldn't just to a link like `<a href="some/mvc/endpoint">`.

> We don't need to make the server smart if we make the client explicit, and it also makes it easier to reason about. The common case/pattern implemented by most UI frameworks is that your SPA takes over the url space once it has been loaded on the client.

That's what I thought originally. But in practice, people had no way to realise there were extra steps to follow, which is why we changed it to be automatic.

I'm not aware of any major drawbacks with it being automatic via the `IsNavigationIntercepted` mechanism. That's an internal implementation detail (in that only `blazor.*.js` code invokes `NotifyLocationChanged`). If it's only a tradeoff between the complexity of this implementation detail and the developer experience, I'd side with the developer experience in this case, especially because we've already tried it the other way around.

Reply:
Fair enough, I remember now when we changed this. I was trying to see if we can reduce the public API to the minimum.

Reply:
I have no plans to change any of this, and it sounds like you both agree now.